### PR TITLE
Kafka: Settings and KRaft support

### PR DESCRIPTION
**Pull request by srhb**
###### Description of changes

Motivation: Support more possible configurations, specifically KRaft (and various migration modes) without having to bail out and write server.properties out by hand.

- [x] Introduce and migrate options to renderable `settings`
- [x] Drop default for apache-kafka user homedir which will become hard to keep sensible once log dirs are split into metadata and data log dirs.
- [x] Try to ensure users migrate safely with assertions and dropped default for log.dirs
- [x] Write release note
- [x] Rewrite NixOS test to support KRaft

Discussion points:

I wrote this module by converting our existing in-house clusters (we have both KRaft and zookeeper clusters) and trying to find a sensible set of options that works for both. While doing this, I needed config references across our deployments, which is why I chose `lazyAttrsOf`. I could have written out more options by hand, but frankly it seems completely arbitrary which options users might rely on references. Let me know if this is a poor choice.

I chose to "shoehorn" in a dropped default for logDirs. I found the existing default scary, and since users are presented with warnings/errors and hopefully a read of the release notes, now seems like a good time to make them actively deliberate their values.

I dropped the homedir for the apache-kafka users since, to my knowledge, it is never used, and the value choice of "first logdir" seems pretty arbitrary to me. /var/empty should be the new default, and it's readonly.

KRaft has some rather annoying requirements of "formatting" log dirs (provisioning certain metadata files) out-of-band of the server process so I've introduced regular NixOS options to support doing this automatically in `preStart`. I'm hoping to engage upstream to get rid of this requirement, which seems like a step in the wrong direction for Kafka, but here we are. See [KIP-785](https://cwiki.apache.org/confluence/display/KAFKA/KIP-785%3A+Automatic+storage+formatting) and related discussion.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by srhb** [comment]
Sort of fixes https://github.com/NixOS/nixpkgs/issues/203987 in a very un-opinionated way (by letting the settings be agnostic to the type of cluster, for now)

**Comment by srhb** [comment]
@ofborg test kafka.kafka kafka.kafka_kraft

File: nixos/modules/services/misc/apache-kafka.nix
Comment (roberth) [inline nixos/modules/services/misc/apache-kafka.nix:172]: This escaping shouldn't be necessary.

```suggestion
      [ "services" "apache-kafka" "settings" "broker.id" ])
```

If anything tries to parse that `.` or renders this as an option path without escaping, that's its mistake, not a mistake here.

**Review by roberth — COMMENTED** [review]
I can only review the configuration-y bits; don't know enough about kafka.

File: nixos/modules/services/misc/apache-kafka.nix
Comment (srhb) [reply nixos/modules/services/misc/apache-kafka.nix]: Makes sense! Done, and `serverProperties` now defaults to the empty string instead of `null` for convenience here.

File: nixos/modules/services/misc/apache-kafka.nix
Comment (srhb) [reply nixos/modules/services/misc/apache-kafka.nix]: Understood, removed :)

**Comment by srhb** [comment]
@roberth Thanks! That's exactly what I need. :smile:

**Comment by srhb** [comment]
https://github.com/ofborg test kafka.kafka kafka.kafka_kraft

**Review by eskytthe — APPROVED** [review]
(Note: I work in the same team as Sarah)

Looking good. Have tested different configs -   works fine. 

´mkRenamedOptionModule´ part works fine - ala:
`trace: warning: The option `services.apache-kafka.brokerId' defined in `/home/ers/work/deployments/deployments/kafka/kafka.nix' has been renamed to `services.apache-kafka.settings."\"broker.id\""'.`

Running tests from  `nixpkgs-review pr 224611` works fine.

Here at DBC we have tested the module with both zookeeper and kraft kafka clusters - works fine. 

Approved. 

Note: New options from manual:
`services.apache-kafka.enable

    Whether to enable Apache Kafka event streaming broker.

    Type: boolean

    Default: false

    Example: true

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.package

    The kafka package to use

    Type: package

    Default: pkgs.apacheKafka

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.clusterId

    KRaft mode ClusterId used for formatting log directories. Can be generated with kafka-storage.sh random-uuid

    Type: null or string

    Default: null

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.formatLogDirs

    Whether to format log dirs in KRaft mode if all log dirs are unformatted, ie. they contain no meta.properties.

    Type: boolean

    Default: false

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.formatLogDirsIgnoreFormatted

    Whether to ignore already formatted log dirs when formatting log dirs, instead of failing. Useful when replacing or adding disks.

    Type: boolean

    Default: false

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.jre

    The JRE with which to run Kafka

    Type: package

    Default: pkgs.apacheKafka.passthru.jre

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.jvmOptions

    Extra command line options for the JVM running Kafka.

    Type: list of string

    Default: [ ]

    Example:

    [
      "-Djava.net.preferIPv4Stack=true"
      "-Dcom.sun.management.jmxremote"
      "-Dcom.sun.management.jmxremote.local.only=true"
    ]

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.log4jProperties

    Kafka log4j property configuration.

    Type: strings concatenated with “\n”

    Default:

    ''
      log4j.rootLogger=INFO, stdout
      
      log4j.appender.stdout=org.apache.log4j.ConsoleAppender
      log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
      log4j.appender.stdout.layout.ConversionPattern=[%d] %p %m (%c)%n
    ''

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.serverProperties

    Literal server.properties content. Will be appended to the config after services.apache-kafka.settings

    Type: strings concatenated with “\n”

    Default: ""

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.settings

    Kafka broker configuration server.properties.

    Note that .properties files contain mappings from string to string. Keys with dots are NOT represented by nested attrs in these settings, but instead as quoted strings (ie. settings."broker.id", NOT settings.broker.id).

    Type: lazy attribute set of (null or boolean or signed integer or string or list of (boolean or signed integer or string))

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.settings."broker.id"

    Broker ID. -1 or null to auto-allocate in zookeeper mode.

    Type: null or signed integer

    Default: null

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.settings.listeners

    Kafka Listener List. See listeners.

    Type: list of string

    Default:

    [
      "PLAINTEXT://localhost:9092"
    ]

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> 
services.apache-kafka.settings."log.dirs"

    Log file directories.

    Type: list of path

    Declared by:
    <nixpkgs/nixos/modules/services/misc/apache-kafka.nix> `

**Comment by srhb** [comment]
I'm refreshing this for another.

Last time I failed to get approval from someone who isn't on my team, which felt a bit too nepotism-adjacent for my taste. I'll try once more and/or lower my standards. :sweat_smile:

**Comment by srhb** [comment]
@ofborg test kafka.kafka kafka.kafka_kraft

**Comment by srhb** [comment]
Think this is back on track. I did:

- A minor simplification of the properties rendering logic.
- Moved release notes to 23.11
- Checked that the log dir formatting is unfortunately still necessary -- see https://github.com/apache/kafka/pull/11549 for the fix which sadly hasn't landed. :cry:

I would appreciate a review from someone outside of my org. :)

**Review by roberth — COMMENTED** [review]
Great PR!

Have a few superficial suggestions, but the best thing you could do is to start a section in the manual about Kafka on NixOS.

I have no domain knowledge about Kafka to speak of, so I'll have to trust you to get that part right, and it seems that you have the experience.

**Comment by srhb** [comment]
@roberth Thank you for the thorough review! I've tacked on commits at the end marked "wip" (that I mean to squash in later) to hopefully aid in re-review, and I'll respond to each review thread individually.

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (srhb) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: I have tried to cut back on the release note and I have implemented the new option as you suggested, enforcing a `mkForce` for the `configFiles.serverProperties` by setting it in `config`.

I have added a note to the manual on how to build this, as you said.

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (srhb) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: I've done this, please take a look at the manual section to see if this looks better. It's a bit wordy, but maybe that's fine :sweat:

File: nixos/modules/services/misc/apache-kafka.nix
Comment (srhb) [reply nixos/modules/services/misc/apache-kafka.nix]: No longer relevant as I dropped support for joining anything to `settings` as requested.

File: nixos/modules/services/misc/apache-kafka.nix
Comment (srhb) [reply nixos/modules/services/misc/apache-kafka.nix:106]: Done

File: nixos/modules/services/misc/apache-kafka.nix
Comment (srhb) [reply nixos/modules/services/misc/apache-kafka.nix:175]: I'm not sure I understood this review comment, sorry. Did you notice that each message differs where I have referred to the new `settings`-based option instead? Or what did you intend with the let-binding?

File: nixos/modules/services/misc/apache-kafka.nix
Comment (srhb) [reply nixos/modules/services/misc/apache-kafka.nix]: Yes, done.

**Comment by srhb** [comment]
Gente re-ping. Would love to get approval and get this in before the window closes. C:

**Comment by srhb** [comment]
Well, I did the rebase, hopefully it's still clear how I addressed the review comments.

**Review by roberth — APPROVED** [review]
LGTM. Still relying on your and eskytthe's testing.
TIL this package does have a maintainer, @ragnard, but they [haven't interacted](https://github.com/NixOS/nixpkgs/issues?q=author%3Aragnard+) for years, so I'm not going to further delay these changes.

@srhb Would you like to be listed as maintainer in the package?

**Pull request merged**
